### PR TITLE
TF Examples Rewrite

### LINKED ISSUE #18334: Fail to reproduce tf2 bert-large f1-score on SQuADv1.1
### System Info

transformers=4.20.1, python=3.7, tensorflow-gpu=2.9.1.

### Who can help?

@Rocketknight1 @sgugger @patil-suraj 

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

1. Download bert-large tf2 pretrained models, .h5 file from [huggingface](https://huggingface.co/bert-large-uncased/tree/main).
2. Run [pytorch question-answering example](https://github.com/huggingface/transformers/tree/main/examples/pytorch/question-answering) for reference. My script is:
```
python run_qa.py \
  --model_name_or_path $BERT_DIR \
  --dataset_name $SQUAD_DIR \
  --do_train \
  --do_eval \
  --per_device_train_batch_size 12 \
  --learning_rate 3e-5 \
  --num_train_epochs 2 \
  --max_seq_length 128 \
  --doc_stride 48 \
  --output_dir $OUTPUT \
  --save_steps 10000 \
  --overwrite_cache \

```
I got an f1-score **90.3953%**. Note that the pretrained model are loaded from tf2 .h5 checkpoints.

3. Run [tf2 question-answering example](https://github.com/huggingface/transformers/tree/main/examples/tensorflow/question-answering) with the same setting of pytorch example. My script is:
```
python run_qa.py \
  --model_name_or_path $BERT_DIR \
  --dataset_name $SQUAD_DIR \
  --do_train \
  --do_eval \
  --per_device_train_batch_size 12 \
  --learning_rate 3e-5 \
  --num_train_epochs 2 \
  --max_seq_length 128 \
  --doc_stride 48 \
  --output_dir $OUTPUT \
  --save_steps 10000 \
  --overwrite_cache \
```
I only got an f1-score **88.5672%**, which is much lower than expected and pytorch results (90.3953%).

### Expected behavior

TF2 question-answering example should achieve similar f1-score with results of corresponding pytroch example.
Or you guys could provide script examples that achieves the target F1-score. Thanks.

### PR DESCRIPTION
**Pull request by Rocketknight1**
This PR is a rewrite of the TF examples, including several modern methods. I'm focusing on updating everything to use modern methods like `prepare_tf_dataset` and the `evaluate` library as well as adding features and functionality I missed when I first ported them, since `transformers` TF support was much shakier when these were first written (and I didn't know the library as well).

Just a draft for now, will ping reviewers when it's ready!

TO DO:

- [x] Draft rewrite for all scripts
- [x] Test run all scripts
- [x] Make sure we're handling batch sizes correctly in multi-GPU/TPU scopes
- [x] Make sure we're correctly using AdamW + LR decay everywhere
- [x] Make sure we're using `evaluate` instead of `load_metric`
- [x] Add metadata for `push_to_hub`
- [x] Add explanatory comments for things like `KerasMetricCallback`, `jit_compile` and `PushToHubCallback` where appropriate
- [x] Replace all the old ad-hoc data loading code with `prepare_tf_dataset`
- [x] Add explanatory links to the docs whenever we use `prepare_tf_dataset`
- [x] Add example tests
- [x] Make sure there's no case where we pass `optimizer=None` to `compile()`
- [ ] Final manual testing
- [x] ~Add HF metrics like MaskedAccuracy?~

Fixes #18334

### CODE REVIEW

**Comment by Rocketknight1** [comment]
This is now ready for review @sgugger @gante! I'm tracking down a couple of remaining bugs in the tests and doing some final manual checks, but almost everything should be finished by now.

File: src/transformers/training_args.py [hunk 1]
```diff
@@ -1039,25 +1040,25 @@ def __post_init__(self):
             self.greater_is_better = self.metric_for_best_model not in ["loss", "eval_loss"]
         if self.run_name is None:
             self.run_name = self.output_dir
+        if self.framework == "pt" and is_torch_available():
```
Comment (sgugger) [inline src/transformers/training_args.py:1043]: Let's maybe have a very assert when `self.framwork=="pt"` that `is_torch_available()` is `True`, so we can remove all `is_torch_available()`.

**Review by sgugger — COMMENTED** [review]
Thanks a lot for diving into this! You should also set up the TensorFlow example job in `.circleci/config.yml` so we see it running (and that would make sure you're not breaking anything in this PR!).

File: src/transformers/training_args.py (hunk 1 above)
Comment (Rocketknight1) [reply src/transformers/training_args.py:1043]: That was my plan originally, but unfortunately the Flax examples use `TrainingArguments` directly, there is no `FlaxTrainingArguments`! Do you want me to add `FlaxTrainingArguments` (with `framework = "flax"`) and change the Flax examples to use it? Or should we just use this hack for now and split the `TrainingArguments` in a future PR?

Comment (sgugger) [reply src/transformers/training_args.py:1043]: Ah, hadn't thought of that. Let's leave it as is for now then, but we really need that `BaseTrainingArguments` class ;-)

File: examples/tensorflow/test_tensorflow_examples.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
+# coding=utf-8
+# Copyright 2018 HuggingFace Inc..
```
Comment (gante) [inline examples/tensorflow/test_tensorflow_examples.py:2]:
```suggestion
# Copyright 2022 HuggingFace Inc.
```
:D

File: examples/tensorflow/token-classification/run_ner.py [hunk 3]
```diff
@@ -517,30 +486,83 @@ def compute_metrics():
...
+        # region Preparing push_to_hub and model card
+        push_to_hub_model_id = training_args.push_to_hub_model_id
+        model_name = model_args.model_name_or_path.split("/")[-1]
+        if not push_to_hub_model_id:
+            if data_args.dataset_name is not None:
+                push_to_hub_model_id = f"{model_name}-finetuned-{data_args.dataset_name}"
+            else:
+                push_to_hub_model_id = f"{model_name}-finetuned-token-classification"
+
+        model_card_kwargs = {"finetuned_from": model_args.model_name_or_path, "tasks": "summarization"}
```
Comment (gante) [inline examples/tensorflow/token-classification/run_ner.py:498]:
```suggestion
        model_card_kwargs = {"finetuned_from": model_args.model_name_or_path, "tasks": "token-classification"}
```

File: examples/tensorflow/translation/run_translation.py [hunk 4]
```diff
@@ -541,67 +489,149 @@ def preprocess_function(examples):
...
-            return preds, labels
+        # endregion
 
+        # region Preparing push_to_hub and model card
+        push_to_hub_model_id = training_args.push_to_hub_model_id
+        model_name = model_args.model_name_or_path.split("/")[-1]
+        if not push_to_hub_model_id:
+            push_to_hub_model_id = f"{model_name}-finetuned-{data_args.source_lang}-{data_args.target_lang}"
+
+        model_card_kwargs = {"finetuned_from": model_args.model_name_or_path, "tasks": "summarization"}
```
Comment (gante) [inline examples/tensorflow/translation/run_translation.py:605]:
```suggestion
        model_card_kwargs = {"finetuned_from": model_args.model_name_or_path, "tasks": "translation"}
```

**Review by gante — APPROVED** [review]
This is great, the examples are much clearer now <3 And CI will make our life much easier!

(added a few nits)

File: examples/tensorflow/language-modeling/run_clm.py [hunk 5]
```diff
@@ -465,44 +478,88 @@ def group_texts(examples):
...
+        # region Preparing push_to_hub and model card
+        push_to_hub_model_id = training_args.push_to_hub_model_id
+        model_name = model_args.model_name_or_path.split("/")[-1]
+        if not push_to_hub_model_id:
+            if data_args.dataset_name is not None:
+                push_to_hub_model_id = f"{model_name}-finetuned-{data_args.dataset_name}"
+            else:
+                push_to_hub_model_id = f"{model_name}-finetuned-clm"
+
+        model_card_kwargs = {"finetuned_from": model_args.model_name_or_path, "tasks": "summarization"}
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply examples/tensorflow/language-modeling/run_clm.py:541]: Ah, this is an incorrect copy-paste, will fix!

File: examples/tensorflow/question-answering/run_qa.py [hunk 6]
```diff
@@ -621,71 +627,163 @@ def compute_metrics(p: EvalPrediction):
...
+            # if you'd like to compute metrics every epoch that are too complex to be written as
+            # standard Keras metrics, you can use our KerasMetricCallback. See
+            # https://huggingface.co/docs/transformers/main/en/main_classes/keras_callbacks
+
+            eval_predictions = model.predict(eval_dataset)
+            if isinstance(eval_predictions.start_logits, tf.RaggedTensor):
+                # If predictions are RaggedTensor, we densify them. Since they are logits, padding with 0 is a bad idea!
+                # The reason is that a logit of 0 can often end up as quite a high probability value, sometimes even
+                # the highest probability in a sample. Instead, we use a value of -1000, which ensures that the padding
+                # positions are correctly masked.
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply examples/tensorflow/question-answering/run_qa.py]: Done!

File: examples/tensorflow/language-modeling/run_clm.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply examples/tensorflow/language-modeling/run_clm.py]: Fixed!

**Comment by Rocketknight1** [comment]
@sgugger Tests are now enabled in `config.yml` and everything still looks green!

**Comment by sgugger** [comment]
That might because your new job did not run ;-) 
You need to add at the end [here](https://github.com/huggingface/transformers/blob/d7e2d7b40b1070cddfe878e13705725f49a2cf1f/.circleci/config.yml#L1000) for the one at each commit and [there](https://github.com/huggingface/transformers/blob/d7e2d7b40b1070cddfe878e13705725f49a2cf1f/.circleci/config.yml#L1024) for the nigthly one ;-)

**Comment by Rocketknight1** [comment]
image

**Comment by Rocketknight1** [comment]
@sgugger tests are now actually passing! I had to skip one - it fails because of a known issue with shape inference on small datasets in `to_tf_dataset`. There is a PR to fix that at https://github.com/huggingface/datasets/pull/4763 , we just need to wait for that to be merged before we can re-enable the test!

**Review by sgugger — APPROVED** [review]
Great, thanks a lot for all the work on this!

**Pull request merged**